### PR TITLE
Update zap to latest commit on dev branch

### PR DESCRIPTION
**Pull request by peter-edge**
I was working on writing a new Module for another project that also uses zap, and I noticed that FX currently uses zap on master, which is way behind the latest zap development on dev. I think this was on purpose (ie I think the whole purpose of the master branch on zap is so that FX could develop against a stable zap API), but as you approach 1.0 for FX, I think getting the latest updates would be a really good idea, so that users of FX are using zap as it will be going forward.

This also does a global glide update, but I think this is the only breaking change. It would be a good idea to audit the rest of the dependencies for updates before FX goes to 1.0 as well. If this is merged, it would be nice to just merge zap dev into master, and then have this depend on master again.

A few notes:

- This is a half hack job. I am just learning the workings of FX, and this should be a drop-in replacement, but I worked to get this just passing tests. This could use a fine-tooth comb by one of the FX maintainers, and I left some TODOs in the code as it is now.
- I think LogBuilder could use some general work afterwards, but I know you're all pushing for 1.0 so this is for later.
- Instead of using LogConfiguration, it might be nice to directly use zap's new configuration struct.
- A general wish: I wish FX wasn't so tightly tied to zap, I wish we could have other logging libraries as well (and a lot of users will want this). It would be non-trivial but not too difficult to separate the tight coupling to zap.
- This PR was a pain in the butt, and if you want to merge it, we should go over it NOW and get it in before this gets into rebase world. I'm happy to do a video chat to get this done quickly, or if someone else wants to take it over, go for it.

@sectioneight @akshayjshah @glibsm @anuptalwalkar

### CODE REVIEW

**Comment by akshayjshah** [comment]
I'm happy to take this over and manage the rebasing. I'd much prefer to wait 1-3 days and pull in the RC of zap 1.0.

**Comment by peter-edge** [comment]
Please do, thank you :) I'd like to file a "friend of the court" brief and
note I'd love this merged sooner than later, the project I'm working on
will be a lot easier if I don't have to rewrite the logging parts of FX
integration twice (it's not that bad to have to change a few things in this
project twice, but just annoying)

**Comment by sectioneight** [comment]
This sounds good, but keep in mind that we're releasing UberFx beta a week from tomorrow, so if you want this change merged we'll need it before EOD Thursday so we can end-to-end test before release.

**Comment by peter-edge** [comment]
It's "ready" to merge but Aisha's should give it a once over for sure.

**Comment by peter-edge** [comment]
*Akshay (iOS spell correct)

**Comment by peter-edge** [comment]
@akshayjshah when can we have that? :)

I don't think you need to worry about SugaredLogger for now. We can do two things:

1. Wait to merge. In that case, this PR is going to be in rebase hell and it's just going to be annoying work for someone to do :)
2. Just merge now, merge zap dev into master, continue going along as is (where FX depends on zap master)

Obviously I'm biased :)

File: testutils/in_memory_log.go [hunk 1]
```diff
@@ -21,41 +21,35 @@
...
+	f(
+		zap.New(
+			zapcore.WriterFacility(
+				zapcore.NewJSONEncoder(
+					zapcore.EncoderConfig{
+						MessageKey:    "msg",
+						LevelKey:      "level",
+						CallerKey:     "caller",
+						StacktraceKey: "stacktrace",
+						//EncodeTime:     zapcore.EpochTimeEncoder,
```
Comment (sectioneight) [inline testutils/in_memory_log.go]: Why is this here / commented-out?

File: testutils/in_memory_log.go [hunk 2]
```diff
@@ -21,41 +21,35 @@
...
-	allOpts = append(allOpts, zap.DebugLevel, zap.Output(sink), zap.ErrorOutput(errSink))
-	allOpts = append(allOpts, opts...)
-	f(zap.New(zap.NewJSONEncoder(zap.NoTime()), allOpts...), sink)
+func WithInMemoryLogger(t *testing.T, opts []zap.Option, f func(*zap.Logger, *testutils.Buffer)) {
+	buffer := &testutils.Buffer{}
+	f(
+		zap.New(
+			zapcore.WriterFacility(
+				zapcore.NewJSONEncoder(
+					zapcore.EncoderConfig{
```
Comment (sectioneight) [inline testutils/in_memory_log.go]: Holy nesting, batman. Can we pull this out to a `var` to make this more readable?

File: ulog/builder.go [hunk 3]
```diff
@@ -99,108 +112,104 @@ func (lb *LogBuilder) WithSentryHook(hook *sentry.Hook) *LogBuilder {
...
 	// TODO(glib): document that yaml configuration takes precedence or
 	// make the situation better so yaml overrides only the DSN
 	if lb.logConfig.Sentry != nil {
 		if len(lb.logConfig.Sentry.DSN) > 0 {
 			hook, err := sentry.Configure(*lb.logConfig.Sentry)
-			if err == nil {
-				lb.sentryHook = hook
-			} else {
+			if err != nil {
+				// TODO(pedge): it would be nicer to return the error
```
Comment (sectioneight) [inline ulog/builder.go]: Interesting. Why don't we return an error here?

File: ulog/builder.go [hunk 4]
```diff
@@ -99,108 +112,104 @@ func (lb *LogBuilder) WithSentryHook(hook *sentry.Hook) *LogBuilder {
...
-func (lb *LogBuilder) Configure() zap.Logger {
-	lb.log = lb.defaultLogger()
-	options := lb.zapOptions()
-
-	if lb.logConfig.TextFormatter != nil && *lb.logConfig.TextFormatter {
-		return zap.New(zap.NewTextEncoder(), options...)
+func (lb *LogBuilder) Configure() (*zap.Logger, error) {
+	levelEnabler := zap.DebugLevel
+	if !lb.logConfig.Verbose {
+		// TODO(pedge): not set anymore
```
Comment (sectioneight) [inline ulog/builder.go]: ?

File: ulog/builder.go [hunk 5]
```diff
@@ -99,108 +112,104 @@ func (lb *LogBuilder) WithSentryHook(hook *sentry.Hook) *LogBuilder {
...
+		//lb.log.Info("Setting log level", zap.String("level", lb.logConfig.Level))
```
Comment (sectioneight) [inline ulog/builder.go]: ?

File: ulog/builder_test.go [hunk 6]
```diff
@@ -165,6 +170,8 @@ func TestMetricsHook(t *testing.T) {
 	assert.Equal(t, 1, len(r.Counters))
 }
 
+// TODO(pedge): it's non-trivial to test this now with how zap sets up hooks
+/*
```
Comment (sectioneight) [inline ulog/builder_test.go]: wait, was this already commented-out?

File: ulog/doc.go [hunk 7]
```diff
@@ -21,7 +21,7 @@
 // Package ulog is the Logging package.
 //
 // package ulog provides an API wrapper around the logging library
-// zap (https://github.com/uber-go/zap). package ulog uses the builder pattern
+// zap (https://go.uber.org/zap). package ulog uses the builder pattern
```
Comment (sectioneight) [inline ulog/doc.go]: I like the previous link better, since godoc does'nt link back to github but github links to godoc

Comment (sectioneight) [inline ulog/doc.go]: Also, these files are auto generated (`make gendoc`) from README so if you do want to change it, change it there first

**Comment by sectioneight** [comment]
note to @alsamylkin -- this appears to remove `IsDevelopmentEnv()` which is on your task list. We should ensure that we're still using a text outputter in dev, but your work may already be done :)

**Comment by peter-edge** [comment]
As this is right now, it just has you set the logging config separately for development and production. In development.yaml, I have:

```yaml
logging:
  level: debug
  stdout: true
  textFormatter: true
```

In production.yaml, I have:
```yaml
logging:
  level: info
```

I'm not sure if you want to assume the formatter based on environment, in fact I'd argue against this, and against some more auto-magical stuff happening with the logging configuation..I'll write something up.

Also as I just mentioned, zap lets do you:

```
go run main.go | jq .
```

So you don't always want a text formatter in development anyways.

**Comment by sectioneight** [comment]
Having textFormatter: true seems like the right solution to this.

**Comment by peter-edge** [comment]
Hey, I'm starting to agree with @anuptalwalkar on this - zap is still changing a lot, and this is probably still a decent base, but even now it's getting into rebase land. @akshayjshah let me know what you want to do, I really want to get this in pre-beta but it might need refactoring again, you probably have more context

**Comment by akshayjshah** [comment]
Zap has tagged a release candidate for a stable 1.0 release, so I don't anticipate many more changes (if any). I'll take this over today and get it ready for merge.

Fair warning, though - this will involve backwards-incompatible changes.

**Comment by peter-edge** [comment]
Thank you so much! @akshayjshah I would note that I also wanted to get in a refactor of the logging configuration in fx, to match zap more closely, but we might try to do that later?

**Comment by akshayjshah** [comment]
I'd prefer to get it in before UberFX starts their open beta, which sounds like it's early next week. The changes will be visible to users, and I'd prefer to not break them right after they start using a new framework.

**Comment by sectioneight** [comment]
NOTE: I had to pin zap today in #264 to unblock service creation -- right now, master is not usable for us

**Comment by anuptalwalkar** [comment]
We should close this PR in favor of https://github.com/uber-go/fx/pull/290 I will hold off by tomorrow if you disagree.

**Comment by peter-edge** [comment]
sgtm
